File: ios/Sources/BugReport/QueryEngine.swift
import Foundation

/// Filter engine for log queries — mirrors JS BR.query() and Kotlin QueryEngine.
public class QueryEngine {
    public struct Filter {
        public let minLevel: Int?
        public let cat: String?
        public let page: String?
        public let tag: String?
        public let search: String?
        public let since: Int64?
        public let until: Int64?
        public let limit: Int?

        public init(minLevel: Int? = nil, cat: String? = nil, page: String? = nil,
                    tag: String? = nil, search: String? = nil, since: Int64? = nil,
                    until: Int64? = nil, limit: Int? = nil) {
            self.minLevel = minLevel; self.cat = cat; self.page = page
            self.tag = tag; self.search = search; self.since = since
            self.until = until; self.limit = limit
        }
    }

    public func query(_ logs: [LogEntry], filter: Filter = Filter()) -> [LogEntry] {
        var result = logs

        if let lvl = filter.minLevel { result = result.filter { $0.level >= lvl } }
        if let cat = filter.cat { result = result.filter { $0.cat == cat } }
        if let page = filter.page { result = result.filter { $0.page.contains(page) } }
        if let tag = filter.tag { result = result.filter { $0.tag.contains(tag) } }
        if let search = filter.search {
            let s = search.lowercased()
            result = result.filter {
                $0.msg.lowercased().contains(s) ||
                $0.tag.lowercased().contains(s) ||
                $0.page.lowercased().contains(s) ||
                $0.cat.lowercased().contains(s)
            }
        }
        if let since = filter.since { result = result.filter { $0.ts >= since } }
        if let until = filter.until { result = result.filter { $0.ts <= until } }

        result = result.reversed()
        if let limit = filter.limit { result = Array(result.prefix(limit)) }
        return result
    }
}

File: android/src/main/kotlin/com/bugreport/QueryEngine.kt
package com.bugreport

/**
 * Filter engine for log queries — mirrors JS BR.query().
 */
class QueryEngine {

    data class Filter(
        val minLevel: Int? = null,
        val cat: String? = null,
        val page: String? = null,
        val tag: String? = null,
        val search: String? = null,
        val since: Long? = null,
        val until: Long? = null,
        val limit: Int? = null
    )

    fun query(logs: List<LogEntry>, filter: Filter = Filter()): List<LogEntry> {
        var result = logs.toList()

        filter.minLevel?.let { lvl -> result = result.filter { it.level >= lvl } }
        filter.cat?.let { cat -> result = result.filter { it.cat == cat } }
        filter.page?.let { page -> result = result.filter { it.page.contains(page) } }
        filter.tag?.let { tag -> result = result.filter { it.tag.contains(tag) } }
        filter.search?.let { search ->
            val s = search.lowercase()
            result = result.filter {
                it.msg.lowercase().contains(s) ||
                it.tag.lowercase().contains(s) ||
                it.page.lowercase().contains(s) ||
                it.cat.lowercase().contains(s)
            }
        }
        filter.since?.let { since -> result = result.filter { it.ts >= since } }
        filter.until?.let { until -> result = result.filter { it.ts <= until } }

        result = result.reversed()
        filter.limit?.let { limit -> result = result.take(limit) }
        return result
    }
}
